src/NuGet.Core/NuGet.Configuration/PackageNamespaces/SearchTree.cs

```diff
@@ -0,0 +1,160 @@
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using NuGet.Common;
+
+namespace NuGet.Configuration
+{
+    public class SearchTree
+    {
+        private readonly SearchNode _root;
+        private const int PackageIdMaxLength = 100;
+
+        internal SearchTree(IReadOnlyList<PackageSourceSection> packageSourceSections)
+        {
+            if (packageSourceSections == null)
+            {
+                throw new ArgumentNullException(nameof(packageSourceSections));
+            }
+
+            _root = new SearchNode(null);
+
+            foreach (PackageSourceSection packageSourceSection in packageSourceSections)
+            {
+#pragma warning disable CA1308 // Normalize strings to uppercase
+                var packageSourceKey = packageSourceSection.GetPackageSourceKey().ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+
+                foreach (string nameSpaceId in packageSourceSection.GetNameSpaceIds())
+                {
+                    Add(nameSpaceId, packageSourceKey);
+                }
+            }
+        }
+
+        private void Add(string namespaceId, string packageSourceKey)
+        {
+            SearchNode currentNode = _root;
+
+            if (namespaceId == null)
+            {
+                throw new ArgumentNullException(nameof(namespaceId));
+            }
+
+            // To prevent from unwanted behaviour.
+            if (namespaceId.Length > PackageIdMaxLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(namespaceId));
+            }
+
+#pragma warning disable CA1308 // Normalize strings to uppercase
+            namespaceId = namespaceId.ToLowerInvariant().Trim();
+#pragma warning restore CA1308 // Normalize strings to uppercase
+
+            for (int i = 0; i < namespaceId.Length; i++)
+            {
+                char c = namespaceId[i];
+
+                if (c == '*')
+                {
+                    // break here since seeing * means end of expression.
+                    currentNode.IsGlobbing = true;
+                    break;
+                }
+
+                if (!currentNode.Children.ContainsKey(c))
+                {
+                    currentNode.Children[c] = new SearchNode(currentNode);
+                }
+
+                currentNode = currentNode.Children[c];
+            }
+
+            if (string.IsNullOrEmpty(currentNode.NamespaceId))
+            {
+                currentNode.NamespaceId = namespaceId;
+            }
+
+            if (currentNode.PackageSources == null)
+            {
+                currentNode.PackageSources = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase);
+            }
+
+            currentNode.PackageSources.Add(packageSourceKey);
+        }
+
+        public ConfigNameSpaceLookup Find(string term)
```

Comment:
Do we need SearchTree in addition to the namespace config?

As an API SearchTree effectively communicates what the thing that it does is, but it's not super obvious what in particular it's being used for. 

tldr;
I don't have a problem with having a type for the search tree, but down in the restore code, it'd be simpler if we're only passing down the PackageNamespaceConfiguration.

Reply:
I believe we still need this one. I made this one internal, now users doesn't need to know about existing of this SearchTree. But it'll help with trie functionality unit tests later.